src-self-hosted/c_tokenizer.zig

```diff
@@ -0,0 +1,656 @@
+                .Identifier,
+                .Decimal,
+                .Hex,
+                .Bin,
+                .Octal,
+                .GotZero,
+                .Float,
+                .FloatExp,
+                => {
+                    result.bytes = chars[begin_index..i.*];
+                    return result;
+                },
+                .NumLitIntSuffixU,
+                .NumLitIntSuffixL,
+                .NumLitIntSuffixUL,
+                .NumLitIntSuffixLL,
+                .GotLt,
+                => {
+                    return result;
+                },
+                .CharLit,
+                .OpenComment,
+                .Comment,
+                .CommentStar,
+                .Backslash,
+                .String,
+                .ExpSign,
+                .FloatExpFirst,
+                => return error.TokenizingFailed,
+            }
+        }
+        i.* += 1;
+        switch (state) {
+            .Start => {
+                switch (c) {
+                    ' ', '\t', '\x0B', '\x0C' => {},
+                    '\'' => {
+                        state = .CharLit;
+                        result.id = .CharLit;
+                        begin_index = i.* - 1;
+                    },
+                    '\"' => {
+                        state = .String;
+                        result.id = .StrLit;
+                        begin_index = i.* - 1;
+                    },
+                    '/' => {
+                        state = .OpenComment;
+                    },
+                    '\\' => {
+                        state = .Backslash;
+                    },
+                    '\n', '\r' => {
+                        return result;
+                    },
+                    'a'...'z', 'A'...'Z', '_' => {
+                        state = .Identifier;
+                        result.id = .Identifier;
+                        begin_index = i.* - 1;
+                    },
+                    '1'...'9' => {
+                        state = .Decimal;
+                        result.id = .NumLitInt;
+                        begin_index = i.* - 1;
+                    },
+                    '0' => {
+                        state = .GotZero;
+                        result.id = .NumLitInt;
+                        begin_index = i.* - 1;
+                    },
+                    '.' => {
+                        result.id = .Dot;
+                        return result;
+                    },
+                    '<' => {
+                        result.id = .Lt;
+                        state = .GotLt;
+                    },
+                    '(' => {
+                        result.id = .LParen;
+                        return result;
+                    },
+                    ')' => {
+                        result.id = .RParen;
+                        return result;
+                    },
+                    '*' => {
+                        result.id = .Asterisk;
```

Comment:
Should this get translated to `%*`?

Reply:
As far as the tokenization is concerned, it's an asterisk. How to translate to zig source is handled in a separate abstraction layer.

Reply:
Probably, but since wrapping multiplication is not defined for floats it would  require some type checking.